File: Editor/assets/Scripts/Player/States/PlayerDash.cs
using System;
using System.Linq;
using Wiwa;

namespace Game
{
    using EntityId = System.UInt64;
    public class PlayerDash : PlayerBaseState
    {
        private float dashTimer = 0f;
        Vector3 targetPoint;
        Vector3 dashDirection;
        float dashParticleTimer;
        float maxDashTime = 1.5f;// to secure that the dash dosent go wiiii to the infinite
        public override void EnterState(ref PlayerStateMachine stateMachine, EntityId entityId)
        {
            Console.WriteLine("Player Dash");
            dashTimer = stateMachine.GetCharacter().DashCooldown;
            if (stateMachine.movementInput == Vector3Values.zero)
            {
                dashDirection = Mathf.CalculateForward(ref stateMachine.GetTransform());
            }
            else
            {
                dashDirection = stateMachine.movementInput;
            }

            targetPoint = Mathf.PointAlongDirection(stateMachine.GetTransform().LocalPosition, dashDirection, stateMachine.GetCharacter().DashDistance);

            Animator.Blend("dash", false,0.1f, stateMachine.GetEntity());
            maxDashTime = 1.5f;
        }
        public override void UpdateState(ref PlayerStateMachine stateMachine, EntityId entityId)
        {
            if(Vector3.Distance(targetPoint,stateMachine.GetTransform().LocalPosition) <= 1f )
            {
                stateMachine.SwitchState(ref stateMachine, stateMachine.idle);
                return;
            }

            maxDashTime -= Time.DeltaTime();

            if (maxDashTime <= 0)
            {
                stateMachine.SwitchState(ref stateMachine, stateMachine.idle);
                return;
            }
            stateMachine.velocity += dashDirection * stateMachine.GetCharacter().DashSpeed;

            PhysicsManager.SetLinearVelocity(stateMachine.GetEntity(), stateMachine.velocity);

            //particles
            dashParticleTimer -= Time.DeltaTime();
            if (dashParticleTimer < 0)
            {
                EntityId pe_dash = stateMachine.GetChildByName("PE_Dash");

                ref Transform3D transform = ref stateMachine.GetComponent<Transform3D>(pe_dash);

                transform.Rotation.y = stateMachine.GetTransform().Rotation.y;

               // ParticleEmitterManger.ParticleEmitterPlayBatch(pe_dash);

                dashParticleTimer = 0.1f; //better if random between 2 values
            }
        }
        public override void ExitState(ref PlayerStateMachine stateMachine, EntityId entityId)
        {
            dashDirection = Vector3Values.zero;

            stateMachine.DashEnable = false;
            stateMachine.cooldownTimer = stateMachine.GetCharacter().DashCooldown;

            PhysicsManager.SetLinearVelocity(stateMachine.GetEntity(), Vector3Values.zero);
            Animator.RestartAnimation(stateMachine.GetEntity());
        }
        public override void OnCollisionEnter(EntityId id1, EntityId id2, string str1, string str2)
        {

        }
    }
}

File: Wiwa/src/Wiwa/ecs/systems/game/player/states/PlayerDash.cpp
#include "wipch.h"
#include "PlayerDash.h"

Wiwa::PlayerDash::PlayerDash(PlayerStateMachine* stateMachine, EntityId id)
	: PlayerBaseState(stateMachine, id)
{
	m_DashTimer = 0.f;
	m_TargetPoint = glm::vec3(0.f);
	m_DashDirection = glm::vec3(0.f);
	m_DashTime = 1.5f;
}

Wiwa::PlayerDash::~PlayerDash()
{
}

void Wiwa::PlayerDash::EnterState()
{
	m_DashTimer = m_StateMachine->GetCharacter()->DashCooldown;
	if (!m_StateMachine->CanMove())
	{
		m_DashDirection = Math::CalculateForward(m_StateMachine->GetTransform());
	}
	else
	{
		m_DashDirection = m_StateMachine->GetInput();
	}

	m_TargetPoint = Math::PointAlongDirection(m_StateMachine->GetTransform()->localPosition, m_DashDirection, m_StateMachine->GetCharacter()->DashDistance);

	GetAnimator()->Blend("dash", false, 0.1f);
	m_DashTime = 1.5f;
}

void Wiwa::PlayerDash::UpdateState()
{
	if (Math::Distance(m_TargetPoint, m_StateMachine->GetTransform()->localPosition))
	{
		m_StateMachine->SwitchState(m_StateMachine->m_IdleState);
		return;
	}
	m_DashTime -= Time::GetDeltaTimeSeconds();

	if (m_DashTime <= 0.f)
	{
		m_StateMachine->SwitchState(m_StateMachine->m_IdleState);
		return;
	}
	m_StateMachine->SetVelocity(m_DashDirection * m_StateMachine->GetCharacter()->DashSpeed);

	GetPhysics()->getBody()->velocity = Math::ToBulletVector3(m_StateMachine->GetVelocity());

	// TODO: Particles and audio queue
}

void Wiwa::PlayerDash::ExitState()
{
	m_DashDirection = glm::vec3(0.f);
	m_StateMachine->SetDashEnable(false);
	m_StateMachine->ResetCooldown();

	GetPhysics()->getBody()->velocity = btVector3(0.f, 0.f, 0.f);
	GetAnimator()->Restart();
}

void Wiwa::PlayerDash::OnCollisionEnter(Object* object1, Object* object2)
{
}
